### PR TITLE
openshift/v4.16: Stabilize 4.16.0 spec and add openshift 4.17-experimental spec

### PR DESCRIPTION
**Pull request by yasminvalim**
openshift/v4.16: Stabilize 4.16.0 spec

Stabilize on ignition 3.4, fcos 1.5 and base 0.5 specs

---

openshift/v4.16: Remove GRUB password support

Support is still missing in the MCO.

See: https://issues.redhat.com/browse/MCO-630

---

openshift/v4_17_exp: Add 4.17.0-experimental spec

---

Fixes: https://github.com/coreos/butane/issues/533

### CODE REVIEW

File: config/openshift/v4_16/translate.go [hunk 1]
```diff
@@ -194,21 +194,21 @@ func (c Config) ToIgn3_5Unvalidated(options common.TranslateOptions) (types.Conf
...
 }
 
-// ToConfigBytes translates from a v4.16 Butane config to a v4.16 MachineConfig or a v3.5.0 Ignition config. It returns a report of any errors or
+// ToConfigBytes translates from a v4.16 Butane config to a v4.16 MachineConfig or a v3.4.0 Ignition config. It returns a report of any errors or
 // warnings in the source and resultant config. If the report has fatal errors or it encounters other problems
 // translating, an error is returned.
 func ToConfigBytes(input []byte, options common.TranslateBytesOptions) ([]byte, report.Report, error) {
 	if options.Raw {
-		return cutil.TranslateBytes(input, &Config{}, "ToIgn3_5", options)
+		return cutil.TranslateBytes(input, &Config{}, "ToIgn4", options)
```
Comment (travier) [inline config/openshift/v4_16/translate.go:211]:
```suggestion
		return cutil.TranslateBytes(input, &Config{}, "ToIgn3_4", options)
```

This is likely the issue.

File: docs/config-openshift-v4_16.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+parent: Configuration specifications
+nav_order: 142
+---
+
+# OpenShift Specification v4.16.0
+
+The OpenShift configuration is a YAML document conforming to the following specification, with **_italicized_** entries being optional:
+
+<div id="spec-docs"></div>
+
```
Comment (travier) [inline docs/config-openshift-v4_16.md:14]: Looks like the docs generation did not work here. Will have to be re-run.

File: config/openshift/v4_16/translate.go (hunk 1 above)
Comment (yasminvalim) [reply config/openshift/v4_16/translate.go]: Thanks!

**Comment by travier** [comment]
https://github.com/coreos/butane/blob/main/internal/doc/butane.yaml#L352 > Those need to be updated to list 4.16.0 now as part of the GRUB removal change.

File: config/openshift/v4_17_exp/schema.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.)
+
+package v4_17_exp
+
+import (
+	fcos "github.com/coreos/butane/config/fcos/v1_7_exp"
```
Comment (travier) [inline config/openshift/v4_17_exp/schema.go:18]:
```suggestion
	fcos "github.com/coreos/butane/config/fcos/v1_6_exp"
```

**Comment by travier** [comment]
Found two more "typos" but there are other missing changes (GRUB password support should be there).

One way to make this easier when creating the new experimental version is to copy the files from before we stabilized the 4.16-exp version and then updating 16 to 17, as everything else will stay the same.

**Comment by yasminvalim** [comment]
> https://github.com/coreos/butane/blob/main/internal/doc/butane.yaml#L352 > Those need to be updated to list 4.16.0 now as part of the GRUB removal change.

I updated it. Can you take a look? Thanks!

**Comment by yasminvalim** [comment]
> Found two more "typos" but there are other missing changes (GRUB password support should be there).
> 
> One way to make this easier when creating the new experimental version is to copy the files from before we stabilized the 4.16-exp version and then updating 16 to 17, as everything else will stay the same.

I re-did to avoid mistakes while doing the 4.17.0_exp. I think now is closer to the ideal. Can you take another look and see what is missing? Thanks!

File: docs/upgrading-openshift.md [hunk 4]
```diff
@@ -15,6 +15,14 @@ Occasionally, changes are made to OpenShift Butane configs (those that specify `
 
 {% comment %}
 
+## From Version 4.16.0 to 4.17.0
+
+There are no breaking changes between versions 4.16.0 and 4.17.0 of the `openshift` configuration specification. Any valid 4.16.0 configuration can be updated to a 4.17.0 configuration by changing the version string in the config.
+
+{% endcomment %}
+
+{% comment %}
```
Comment (travier) [inline docs/upgrading-openshift.md:24]: You should remove the comment/endcomment for the 4.15 to 4.16 section

**Comment by travier** [comment]
I think this looks good. One change needed for the release notes and then I'll do a final check/review.

**Comment by yasminvalim** [comment]
> I think this looks good. One change needed for the release notes and then I'll do a final check/review.

I think everything looks okay now. Feel free to review it again and tell me if there is anything I could improve!

**Comment by travier** [comment]
Thanks, this looks complete. I did a minor re-split of the commits but changed nothing to what you did.

So LGTM.

**Review by travier — APPROVED** [review]

**Pull request merged**